pkg/threescale/backend/helpers.go

```diff
@@ -0,0 +1,103 @@
+package backend
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/3scale/3scale-go-client/threescale"
+	"github.com/3scale/3scale-go-client/threescale/api"
+)
+
+// blankRequestFromTransactions is used when we need to gather state from the remote backend
+// it parses the transactions for their auth and augments the request to provide nil metrics
+func blankRequestFromTransaction(transaction api.Transaction) api.Transaction {
+	return api.Transaction{
+		Metrics:   nil,
+		Params:    transaction.Params,
+		Timestamp: transaction.Timestamp,
+	}
+}
+
+func computeAffectedMetrics(app *Application, request threescale.Request) api.Metrics {
+	return request.Transactions[0].Metrics.AddHierarchyToMetrics(app.LastResponse.Hierarchy)
+}
+
+// getApplicationFromResponse builds an Application from a response from 3scale client
+func getApplicationFromResponse(resp *threescale.AuthorizeResult) Application {
+	limitCounter := make(map[string]map[api.Period]*Limit)
+
+	for metric, report := range resp.UsageReports {
+		l := &Limit{
+			current: report.CurrentValue,
+			max:     report.MaxValue,
+		}
+		limitCounter[metric] = make(map[api.Period]*Limit)
+		limitCounter[metric][report.PeriodWindow.Period] = l
+	}
+
+	return Application{
+		LastResponse:     resp,
+		LimitCounter:     limitCounter,
+		UnlimitedCounter: make(map[string]int),
+	}
+}
+
+// getAppIdFromTransaction prioritizes and returns a user key if present, defaulting to app id otherwise
+func getAppIDFromTransaction(transaction api.Transaction) string {
+	appID := transaction.Params.UserKey
+
+	if appID != "" {
+		return appID
+	}
+	return transaction.Params.AppID
+}
+
+func generateCacheKeyFromRequest(request threescale.Request) string {
+	return fmt.Sprintf("%s_%s", request.GetServiceID(), getAppIDFromTransaction(request.Transactions[0]))
+}
+
+// getEmptyAuthRequest is a helper method to return a request suitable for a blanket auth request
+func getEmptyAuthRequest(service api.Service, auth api.ClientAuth, params api.Params) threescale.Request {
+	return threescale.Request{
+		Auth:       auth,
+		Extensions: api.Extensions{api.HierarchyExtension: "1"},
+		Service:    service,
+		Transactions: []api.Transaction{
+			{
+				Params: params,
+			},
+		},
+	}
+}
+
+func parseCacheKey(cacheKey string) (service string, application string, err error) {
+	parsed := strings.Split(cacheKey, "_")
+
+	if len(parsed) != 2 {
+		return service, application, fmt.Errorf("error parsing key")
+	}
+
+	service, application = parsed[0], parsed[1]
+
+	if service == "" || application == "" {
+		return service, application, fmt.Errorf("error parsing key")
+	}
+
+	return service, application, nil
+}
+
+func validateTransactions(transactions []api.Transaction) error {
+	if transactions == nil || len(transactions) == 0 {
+		return fmt.Errorf("transaction must be non-nil and non-empty")
+	}
+	return nil
+}
+
+// newApplication creates a new, empty application with maps initialised
+func newApplication() *Application {
```

Comment:
Apart from creating a new, blank application we could have a constructor receiving the response and limits and reuse it from other places (even in tests), ie. `getApplicationFromResponse()`, wdyt?